nx/lib/nx.ex

```diff
@@ -7842,6 +7842,70 @@ defmodule Nx do
     impl!(tensor).sort(tensor, tensor, axis: axis, comparator: comparator)
   end
 
+  @doc """
+  Solve the equation a x = b for x, assuming a is a triangular matrix.
+
+  ## Examples
+
+      iex> Nx.triangular_solve(Nx.tensor([[3, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]]), Nx.tensor([4, 2, 4, 2]))
+      #Nx.Tensor<
+        f64[4]
+        [1.33333333, -0.66666667, 2.66666667, -1.33333333]
+      >
+
+  ### Error cases
+
+      iex> Nx.triangular_solve(Nx.tensor([[3, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]]), Nx.tensor([4, 2, 4, 2]), trans: 9)
+      ** (ArgumentError) unknown trans 9, expected 0 or 'N' for a x = b
+
+      iex> Nx.triangular_solve(Nx.tensor([[3, 0, 0, 0], [2, 1, 0, 0]]), Nx.tensor([4, 2, 4, 2]), trans: 0)
+      ** (ArgumentError) expected a square matrix, got: {2, 4}
+
+      iex> Nx.triangular_solve(Nx.tensor([[3, 0, 0, 0], [2, 1, 0, 0], [1, 1, 1, 1], [1, 1, 1, 1]]), Nx.tensor([4]), trans: 0)
+      ** (ArgumentError) incompatible dimensions
+  """
+  @doc type: :linalg
+  def triangular_solve(a, b, opts \\ []) do
+    %T{shape: s1 = {m, n}} = a = tensor!(a)
+    %T{shape: {q}} = b = tensor!(b)
+
+    case shape(s1) do
+      {n, n} -> {n, n}
+      other -> raise ArgumentError, "expected a square matrix, got: #{inspect(other)}"
+    end
+
+    if m != q, do: raise(ArgumentError, "incompatible dimensions")
+
+    assert_keys!(opts, [:trans])
```

Comment:
I think Nx's pattern is to have more extensive names instead of abbreviations for the options

Reply:
@polvalente I was going for a something like that available in scipy. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_triangular.html

What do you think of following that route?

Reply:
Sounds good. I'd rename this to something more descriptive like `system_type`, and only accepting either atoms or numbers as the values. @josevalim might be able to guide us towards the best solution here.

Reply:
I would maybe hold off on adding this until we add support for complex numbers. It's supposed to represent `transpose` with the options of no transpose, transpose, and conjugate transpose, correct? In that case, is it the same to just do: `triangular_solve(transpose(a))`? If the results are the same, then I would hold off on adding it for the time being and then we can revisit with complex numbers.

Reply:
Seems reasonable. Until we have complex numbers there's really not much of a difference (even with respect to implementation, where some algorithms change a bit)

Reply:
@seanmor5 I believe that setting trans to 0 or 'N' is the default. So I am only dealing with the default case for when the system a x = b

No other cases are supported.